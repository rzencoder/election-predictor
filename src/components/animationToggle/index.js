import "./animationToggle.scss";

export default function AnimationToggle({ setAnimations, animations }) {
  return (
    <div className="animations">
      <div className="animations-text">Animations</div>
      <label className="switch">
        <input
          checked={animations}
          type="checkbox"
          aria-label="toggle-animations"
          onChange={() => setAnimations(!animations)}
        />
        <span className="slider round"></span>
      </label>
    </div>
  );
}
