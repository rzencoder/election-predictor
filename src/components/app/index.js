import { useState, useEffect } from "react";
import {
  Districts,
  Header,
  Map,
  SmallStates,
  VotesBar,
  ShareLink,
  Tooltip,
  AnimationToggle,
} from "../../components";
import states from "../../data/states.json";
import { parseURLMapData } from "../../utils";
import ReactTooltip from "react-tooltip";
import { getPresident } from "../../utils";
import "./app.scss";
import YearSelect from "../yearSelect";

export default function App() {
  const [stateData, setStateData] = useState(states);
  const [animations, setAnimations] = useState(true);
  const [tooltipContent, setTooltipContent] = useState(null);
  const [year, setYear] = useState("Predictor");

  // Check if map url parameter and parse it into data to display in the map
  useEffect(() => {
    const queryString = window.location.search;
    if (queryString) {
      const urlParams = new URLSearchParams(queryString);
      const mapParameter = urlParams.get("map");
      // Checking if 19 as data stored in base 3. 50 states + 6 districts + blank
      if (mapParameter.length === 19) {
        setTimeout(() => {
          const mapString = parseURLMapData(mapParameter);
          const mapData = stateData.map((el, index) => {
            return { ...el, party: parseInt(mapString[index]) };
          });
          setStateData(mapData);
          setAnimations(false);
        }, 200);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // Change states data when user changes the election year
  useEffect(() => {
    if (year !== "Predictor") {
      setAnimations(false);
    } else {
      setAnimations(true);
    }
    // If user selects the predictor reset the data to the original
    if (year === "Predictor") return setStateData(states);
    const data = stateData.map((el) => {
      const index = el.history.findIndex((el) => el.year === parseInt(year));
      const party = el.history[index].party;
      const votes = el.history[index].votes;
      return { ...el, party, votes };
    });
    setStateData(data);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [year]);

  return (
    <main>
      <Header />
      <VotesBar stateData={stateData} year={year} animations={animations} />
      <section className="options-container">
        <div>
          <YearSelect yearData={stateData[0]} setYear={setYear} year={year} />
          <AnimationToggle
            setAnimations={setAnimations}
            animations={animations}
          />
        </div>
        {year !== "Predictor" && (
          <div className="title-historical-result">{`${year} ${getPresident(
            parseInt(year),
            1
          )} vs ${getPresident(parseInt(year), 2)}`}</div>
        )}
      </section>

      <section className="container">
        <Map
          stateData={stateData}
          setStateData={setStateData}
          setTooltipContent={setTooltipContent}
        />
        <ReactTooltip className="tooltip-container">
          {tooltipContent && (
            <Tooltip
              matchingState={stateData.find((s) => s.val === tooltipContent)}
            />
          )}
        </ReactTooltip>
        <aside className="states-panel">
          <SmallStates stateData={stateData} setStateData={setStateData} />
          <Districts stateData={stateData} setStateData={setStateData} />
        </aside>
      </section>
      <ShareLink stateData={stateData} />
    </main>
  );
}
