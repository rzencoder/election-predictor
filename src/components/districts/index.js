import { getStateColor } from "../../utils";
import districts from "../../data/districts.json";
import { memo } from "react";
import "./districts.scss";

function Districts({ stateData, setStateData }) {
  // Handle clicking on states too small to click directly on the map
  const handleSmallStateClick = (id) => {
    const newStateData = stateData.map((el) => {
      // Change party when state selected and roll back to a blank if neither rep or dem.
      const partyValue = el.party === 2 ? 0 : el.party + 1;
      return el.id === id ? { ...el, party: partyValue } : el;
    });
    setStateData(newStateData);
  };

  // Render a table row for Maine and Nebraska
  function renderStateDistricts(stateDistricts) {
    return (
      <tr>
        <td className="district-state-name">{stateDistricts[0]}</td>
        {stateDistricts.map((el) => {
          const matchingState = stateData.find((s) => s.id === el);
          const background = getStateColor(matchingState.party);
          return (
            <td
              key={`${matchingState.id}-district`}
              className={`district-votes ${background}`}
            >
              <button
                onClick={() => {
                  handleSmallStateClick(el);
                }}
              >
                {matchingState.id === "NE"
                  ? 2
                  : matchingState.id === "ME"
                  ? 2
                  : matchingState.votes}
              </button>
            </td>
          );
        })}
      </tr>
    );
  }

  return (
    <div className="districts-container">
      <table>
        <thead>
          <tr className="table-headers">
            <td></td>
            <td rowSpan="2">state</td>
            <td colSpan="3">districts</td>
          </tr>
          <tr className="district-numbers">
            <td></td>
            <td>1</td>
            <td>2</td>
            <td>3</td>
          </tr>
        </thead>
        <tbody>
          {renderStateDistricts(districts.nebraska)}
          {renderStateDistricts(districts.maine)}
        </tbody>
      </table>
    </div>
  );
}

export default memo(Districts);
