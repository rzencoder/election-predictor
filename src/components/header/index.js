import "./header.scss";

export default function Header() {
  return (
    <header>
      <h1 className="header-title">
        <span className="title-span">US</span> Presidential{" "}
        <span className="title-span">Election</span>
      </h1>
      <div className="header-subtitle">
        <span className="header-star red-star">★</span>
        <span className="header-star blue-star">★</span>
        <span className="header-star red-star">★</span>
        <span className="header-subtitle-text">Predictor</span>
        <span className="header-star red-star">★</span>
        <span className="header-star blue-star">★</span>
        <span className="header-star red-star">★</span>
      </div>
    </header>
  );
}
