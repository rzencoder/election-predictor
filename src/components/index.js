export { default as Header } from "./header";
export { default as Map } from "./map";
export { default as SmallStates } from "./smallStates";
export { default as SVGDefs } from "./svgDefs";
export { default as VotesBar } from "./votesBar";
export { default as ShareLink } from "./shareLink";
export { default as Districts } from "./districts";
export { default as AnimationToggle } from "./animationToggle";
