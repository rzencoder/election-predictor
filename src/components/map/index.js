import { geoCentroid } from "d3-geo";
import {
  ComposableMap,
  Geographies,
  Geography,
  Marker,
} from "react-simple-maps";
import { SVGDefs } from "../index.js";
import geoData from "../../data/geo.json";
import smallStatesData from "../../data/smallStates.json";
import { handleStateColor } from "../../utils";
import "./map.scss";

export default function Map({ stateData, setStateData }) {
  const handleStateClick = (id) => {
    const newStateData = stateData.map((obj) => {
      // Dem: 1, Rep: 2, Blank: 0
      let partyValue = obj.party === 2 ? 0 : obj.party + 1;
      // Find matching state from stateDate and assign party change
      return obj.val === id ? { ...obj, party: partyValue } : obj;
    });
    // Change Nebraska districts to state party on state click
    if (id === "31") {
      newStateData[51].party = newStateData[27].party;
      newStateData[52].party = newStateData[27].party;
      newStateData[53].party = newStateData[27].party;
    } else if (id === "23") {
      // Change Maine districts to state party on state click
      newStateData[54].party = newStateData[19].party;
      newStateData[55].party = newStateData[19].party;
    }
    setStateData(newStateData);
  };

  return (
    <div className="map-container">
      <ComposableMap projection="geoAlbersUsa">
        <SVGDefs stateData={stateData} />
        <Geographies geography={geoData}>
          {({ geographies }) => (
            <>
              {geographies.map((geo) => {
                const fill = handleStateColor(geo, stateData);
                return (
                  <Geography
                    key={geo.rsmKey}
                    stroke="#FFF"
                    geography={geo}
                    onClick={() => handleStateClick(geo.id)}
                    fill={fill}
                  ></Geography>
                );
              })}
              {geographies.map((geo) => {
                const centroid = geoCentroid(geo);
                const currentState = stateData.find((el) => el.val === geo.id);
                if (smallStatesData.includes(currentState.id)) {
                  return;
                }
                return (
                  <g key={geo.rsmKey + "-state"}>
                    {currentState && centroid[0] > -160 && centroid[0] < -67 && (
                      <Marker coordinates={centroid}>
                        <text
                          y="2"
                          onClick={() => handleStateClick(geo.id)}
                          fontSize={13}
                          textAnchor="middle"
                          fill={currentState.id === "HI" ? "#000" : "#fff"}
                        >
                          {currentState.id}
                        </text>
                        <text
                          y="16"
                          onClick={() => handleStateClick(geo.id)}
                          fontSize={11}
                          textAnchor="middle"
                          fill={currentState.id === "HI" ? "#000" : "#fff"}
                        >
                          {currentState.votes}
                        </text>
                      </Marker>
                    )}
                  </g>
                );
              })}
            </>
          )}
        </Geographies>
      </ComposableMap>
    </div>
  );
}
