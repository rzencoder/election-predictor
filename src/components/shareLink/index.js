import { useState, useEffect } from "react";
import { convertMapData } from "../../utils";
import { CopyToClipboard } from "react-copy-to-clipboard";
import copyIcon from "./copy-icon.png";
import { memo } from "react";
import "./shareLink.scss";

function ShareLink({ stateData }) {
  const [showLink, setShowLink] = useState(false);
  const [link, setLink] = useState("");
  const [copied, setCopied] = useState(false); // eslint-disable-line no-unused-vars

  const linkURL = process.env.REACT_APP_LINK_URL || "http://localhost:3000";

  // Convert state data into a url parameter for sharing on selection
  useEffect(() => {
    if (showLink) {
      setLink(convertMapData(stateData));
      setCopied(false);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [showLink]);

  // Hide link when user changes map
  useEffect(() => {
    setShowLink(false);
    setLink("");
  }, [stateData]);

  return (
    <section className="share-link-container">
      {!showLink && (
        <button
          className="share-link-button"
          data-testid="share-link-button"
          onClick={() => setShowLink(!showLink)}
        >
          {"</> "}Share Map
        </button>
      )}
      {showLink && (
        <div className="link-copy">
          <div className="link-text">{`${linkURL}?map=${link}`}</div>
          <CopyToClipboard
            text={`${linkURL}?map=${link}`}
            onCopy={() => setCopied(true)}
          >
            <button className="copy-button">
              <img src={copyIcon} alt="" />
              Copy
            </button>
          </CopyToClipboard>
        </div>
      )}
    </section>
  );
}

export default memo(ShareLink);
