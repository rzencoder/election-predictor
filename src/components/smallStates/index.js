import smallStatesData from "../../data/smallStates.json";
import { memo } from "react";
import { getStateColor } from "../../utils";
import "./smallStates.scss";

function SmallStates({ stateData, setStateData }) {
  // Handle clicking on states too small to click directly on the map
  const handleSmallStateClick = (id) => {
    const newStateData = stateData.map((el) => {
      // Change party when state selected and roll back to a blank if neither rep or dem.
      const partyValue = el.party === 2 ? 0 : el.party + 1;
      return el.id === id ? { ...el, party: partyValue } : el;
    });
    setStateData(newStateData);
  };

  if (!stateData) return null;

  return (
    <div className="small-states-container">
      {smallStatesData.map((el) => {
        const matchingState = stateData.find((s) => s.id === el);
        const background = getStateColor(matchingState.party);
        return (
          <div key={`${matchingState.id}-state`} className="button-container">
            <button
              className={`small-state-button ${background}`}
              id={`${matchingState.id}-small-state`}
              onClick={() => {
                handleSmallStateClick(el);
              }}
            >
              {matchingState.votes}
            </button>
            <div className="small-state-name">
              <div>{el}</div>
            </div>
          </div>
        );
      })}
    </div>
  );
}

export default memo(SmallStates);
