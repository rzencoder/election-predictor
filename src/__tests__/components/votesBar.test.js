import { render } from "@testing-library/react";
import { VotesBar } from "../../components";

const data = [
  {
    id: "FL",
    val: "12",
    votes: 2,
    party: 0,
    name: "Florida",
    history: [
      { year: 2020, party: 2, votes: 29 },
      { year: 2016, party: 2, votes: 29 },
      { year: 2012, party: 1, votes: 29 },
      { year: 2008, party: 1, votes: 27 },
      { year: 2004, party: 2, votes: 27 },
      { year: 2000, party: 2, votes: 25 },
    ],
  },
  {
    id: "GA",
    val: "13",
    votes: 270,
    party: 2,
    name: "Georgia",
    history: [
      { year: 2020, party: 1, votes: 16 },
      { year: 2016, party: 2, votes: 16 },
      { year: 2012, party: 2, votes: 16 },
      { year: 2008, party: 2, votes: 15 },
      { year: 2004, party: 2, votes: 15 },
      { year: 2000, party: 2, votes: 13 },
    ],
  },
  {
    id: "HI",
    val: "15",
    votes: 3,
    party: 1,
    name: "Hawaii",
    history: [
      { year: 2020, party: 1, votes: 4 },
      { year: 2016, party: 1, votes: 4 },
      { year: 2012, party: 1, votes: 4 },
      { year: 2008, party: 1, votes: 4 },
      { year: 2004, party: 1, votes: 4 },
      { year: 2000, party: 1, votes: 4 },
    ],
  },
  {
    id: "ME",
    val: "23",
    votes: 4,
    party: 1,
    name: "Maine",
    history: [
      { year: 2020, party: 1, votes: 4 },
      { year: 2016, party: 1, votes: 4 },
      { year: 2012, party: 1, votes: 4 },
      { year: 2008, party: 1, votes: 4 },
      { year: 2004, party: 1, votes: 4 },
      { year: 2000, party: 1, votes: 4 },
    ],
  },
  {
    id: "NE",
    val: "31",
    votes: 5,
    party: 2,
    name: "Nebraska",
    history: [
      { year: 2020, party: 2, votes: 5 },
      { year: 2016, party: 2, votes: 5 },
      { year: 2012, party: 2, votes: 5 },
      { year: 2008, party: 2, votes: 5 },
      { year: 2004, party: 2, votes: 5 },
      { year: 2000, party: 2, votes: 5 },
    ],
  },
];

const demData = [
  {
    id: "FL",
    val: "12",
    votes: 2,
    party: 0,
    name: "Florida",
    history: [
      { year: 2020, party: 2, votes: 29 },
      { year: 2016, party: 2, votes: 29 },
      { year: 2012, party: 1, votes: 29 },
      { year: 2008, party: 1, votes: 27 },
      { year: 2004, party: 2, votes: 27 },
      { year: 2000, party: 2, votes: 25 },
    ],
  },
  {
    id: "GA",
    val: "13",
    votes: 270,
    party: 1,
    name: "Georgia",
    history: [
      { year: 2020, party: 1, votes: 16 },
      { year: 2016, party: 2, votes: 16 },
      { year: 2012, party: 2, votes: 16 },
      { year: 2008, party: 2, votes: 15 },
      { year: 2004, party: 2, votes: 15 },
      { year: 2000, party: 2, votes: 13 },
    ],
  },
];

describe("<VotesBar />", () => {
  it("renders the <VotesBar />", () => {
    const { container, getByText, queryByText, getByTestId } = render(
      <VotesBar year={"Predictor"} animations={false} stateData={data} />
    );
    expect(getByTestId("votes-bar")).toBeTruthy();
    expect(getByText("Republicans Win")).toBeTruthy();
    expect(getByText("Democrats")).toBeTruthy();
    expect(queryByText("Trump")).toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("displays democrats as winners with enough votes", () => {
    const { getByText, queryByText } = render(
      <VotesBar year={"Predictor"} animations={false} stateData={demData} />
    );
    expect(queryByText("Republicans Win")).toBeFalsy();
    expect(getByText("Republicans")).toBeTruthy();
    expect(getByText("Democrats Win")).toBeTruthy();
    expect(queryByText("Trump")).toBeNull();
  });

  it("renders nothing with no data", () => {
    const { queryByTestId } = render(
      <VotesBar year={null} animations={false} stateData={null} />
    );
    expect(queryByTestId("votes-bar")).toBeNull();
  });

  it("renders different presidents with different year", () => {
    const { getByText, queryByText } = render(
      <VotesBar year={2016} animations={false} stateData={data} />
    );
    expect(getByText("Trump Wins")).toBeTruthy();
    expect(getByText("Clinton")).toBeTruthy();
    expect(queryByText("Biden")).toBeNull();
    expect(queryByText("Republicans")).toBeNull();
    expect(queryByText("Democrats")).toBeNull();
  });

  // it("renders confetti with winner and animations on", () => {
  //   const { getByText, queryByText } = render(
  //     <VotesBar
  //       year={"Predictor"}
  //       animations={true}
  //       stateData={data}
  //     />
  //   );
  //   expect(getByText("Republicans Win")).toBeTruthy();
  // });
});
