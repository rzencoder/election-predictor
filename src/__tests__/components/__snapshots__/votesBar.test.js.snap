// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VotesBar /> renders the <VotesBar /> 1`] = `
<section
  class="bar-container"
  data-testid="votes-bar"
>
  <div
    class="bar-markers"
  >
    <div
      class="markers marker-top"
    />
    <div
      class="marker-divider"
    />
    <div
      class="markers marker-bottom"
    />
  </div>
  <div
    class="bar-labels"
  >
    <div>
      Democrats
       
    </div>
    <div
      class="win"
    >
      Republicans
       
      Win
    </div>
  </div>
  <div
    class="bar"
  >
    <div
      class="party-bar dem-bar false"
      style="width: 0.929368029739777%;"
    >
      <span>
        0
      </span>
    </div>
    <div
      class="swing-bar"
      style="width: 0.37174721189591076%;"
    />
    <div
      class="party-bar rep-bar win"
      style="width: 50.55762081784386%;"
    >
      <span>
        0
      </span>
    </div>
  </div>
</section>
`;
